clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1727,147 +1716,123 @@ class SyclKernelBodyCreator : public SyclKernelFieldHandler {
+    // TODO, better way to detect that we're in an array?
+    SemaRef.getASTContext().hasSameType(FD->getType(), Ty);
+    return FD->getType() != Ty;
   }
 
-  int getDims() {
-    int Dims = 0;
-    for (int i = MemberExprBases.size() - 1; i >= 0; --i) {
-      if (!isa<ArraySubscriptExpr>(MemberExprBases[i]))
-        break;
-      ++Dims;
-    }
-    return Dims;
-  }
-
-  int64_t getArrayIndex(int Idx) {
-    ArraySubscriptExpr *LastArrayRef =
-        cast<ArraySubscriptExpr>(MemberExprBases[Idx]);
-    Expr *LastIdx = LastArrayRef->getIdx();
-    llvm::APSInt Result;
-    SemaRef.VerifyIntegerConstantExpression(LastIdx, &Result);
-    return Result.getExtValue();
-  }
-
-  void createExprForScalarElement(FieldDecl *FD) {
-    llvm::SmallVector<InitializedEntity, 4> InitEntities;
-
-    // For multi-dimensional arrays, an initialized entity needs to be
-    // generated for each 'dimension'. For example, the initialized entity
-    // for s.array[x][y][z] is constructed using initialized entities for
-    // s.array[x][y], s.array[x] and s.array. InitEntities is used to maintain
-    // this.
-    InitializedEntity Entity =
-        InitializedEntity::InitializeMember(FD, &VarEntity);
-    InitEntities.push_back(Entity);
-
-    // Calculate dimension using ArraySubscriptExpressions in MemberExprBases.
-    // Each dimension has an ArraySubscriptExpression (maintains index)
-    // in MemberExprBases. For example, if we are currently handling element
-    // a[0][0][1], the top of stack entries are ArraySubscriptExpressions for
-    // indices 0,0 and 1, with 1 on top.
-    int Dims = getDims();
-
-    // MemberExprBasesIdx is used to get the index of each dimension, in correct
-    // order, from MemberExprBases. For example for a[0][0][1], getArrayIndex
-    // will return 0, 0 and then 1.
-    int MemberExprBasesIdx = MemberExprBases.size() - Dims;
-    for (int I = 0; I < Dims; ++I) {
-      InitializedEntity NewEntity = InitializedEntity::InitializeElement(
-          SemaRef.getASTContext(), getArrayIndex(MemberExprBasesIdx),
-          InitEntities.back());
-      InitEntities.push_back(NewEntity);
-      ++MemberExprBasesIdx;
-    }
+  // Creates an initialized entity for a field/item. In the case where this is a
+  // field, returns a normal member initializer, if we're in a sub-array of a MD
+  // array, returns an element initializer.
+  InitializedEntity getFieldEntity(FieldDecl *FD, QualType Ty) {
+    if (IsArrayElement(FD, Ty))
+      return InitializedEntity::InitializeElement(SemaRef.getASTContext(),
+                                                  ArrayInfos.back().second,
+                                                  ArrayInfos.back().first);
+    return InitializedEntity::InitializeMember(FD, &VarEntity);
+  }
 
+  // TODO: This name seems outdated now :)
+  void addFieldInit(FieldDecl *FD, QualType Ty, MultiExprArg ParamRef) {
+    // TODO: Why is this by copy rather than 'forinit' or value init?
```

Comment:
Copy ends up being more generic, the each have specific issues based on what is coming in.  We could potentially optimize the init at one point by making this one of the other two, but it would have to be on a per-type basis.